File: backend/src/main/java/be/ap/backend/entity/Loan.java
package be.ap.backend.entity;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "loan")
@Data
@NoArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Loan {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @EqualsAndHashCode.Include
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne
    @JoinColumn(name = "location_id", nullable = false)
    private Location location;

    @Column(name = "extended", nullable = false)
    private Byte extended = 0;

    @Column(name = "start", nullable = false)
    private LocalDate start;

    @Column(name = "end", nullable = false)
    private LocalDate end;

    @Column(name = "note", length = 255)
    private String note;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private LoanStatus status = LoanStatus.REQUESTED;

    @Column(name = "created", nullable = false, updatable = false)
    private LocalDateTime created = LocalDateTime.now();

    @Column(name = "closed", nullable = false)
    private Boolean closed = false;

    @Column(name = "returned_at")
    private LocalDate returnedAt;

    @Column(name = "notified", nullable = false)
    private Boolean notified = false;

    @OneToMany(mappedBy = "loan", cascade = CascadeType.ALL)
    private Set<LoanBook> loanBooks = new HashSet<>();

    @Column(name = "group_id")
    private String groupId;
}

File: frontend/src/app/models/loan.ts
import { CreateLoanBookDTO, LoanBookDTO } from './loanBook';

export enum LoanStatus {
  REQUESTED = 'REQUESTED',
  RECEIVED = 'RECEIVED',
  RETURNED = 'RETURNED',
  ACCEPTED = 'ACCEPTED',
  DECLINED = 'DECLINED',
}

export interface CreateLoanDTO {
  userId: number;
  locationId: number;
  extended: number;
  start: string;
  end: string;
  note: string;
  status: LoanStatus;
  closed: boolean;
  books: CreateLoanBookDTO[];
}

export interface LoanDTO {
  id: number;
  username: String;
  locationId: number;
  extended: number;
  start: string;
  end: string;
  note: string;
  status: LoanStatus;
  closed: boolean;
  created: string;
  books: LoanBookDTO[];
  groupId?: string;
  extendPeriod: number;
}
